### ISSUE TITLE
Properly truncate and randomly load balance large answers

### ISSUE THREAD

**Issue by jdef** [mesosphere/mesos-dns#237]
wondering if we need some kind of max-record-count that limits the number if IPs returned for A records, or addresses returned for SRVs in large clusters (10k). for example, the "slave.mesos" name that maps to all slave IPs in the cluster probably doesn't scale very well to large cluster sizes. same goes for 'task.framework.name' for very large numbers of tasks with the same name.

what's the value in getting back 10k addresses or 10k IPs?

**Comment by tsenart** [comment]
@jdef: I think our answers should be limited to what fits in a DNS UDP datagram without truncation. I don't think we need a limit the number of IPs in the server memory, only in DNS answers. We should randomize the which IPs are included in the answer.

**Comment by discordianfish** [comment]
@tsenart But this would break for example the prometheus DNS based SD. That expects that we return *all* slaves.

**Comment by tsenart** [comment]
@discordianfish: We need to do this regardless of whether we support [EDNS(0)](https://tools.ietf.org/html/rfc6891), but that should alleviate the problem for Prometheus, as long as it has EDNS(0) enabled DNS client. The upper bound of records we can have in the answer section is 65535 due to the `ANCOUNT` field in the header being an `uint16`. If the client won't support EDNS(0) then it could try over TCP, which is done by default in the native Go DNS client if the returned message over UDP is truncated. Over TCP, we could always serve up to 65535 answers in a single message.

**Comment by discordianfish** [comment]
@tsenart Yes, that's what we should do but you said above that we should always limit the number so it fits into UDP. Instead we should return all, even if this means we need to truncate and have clients fall back to tcp.

**Comment by tsenart** [comment]
@discordianfish: By definition, when you truncate, you can't return all :-)

**Comment by discordianfish** [comment]
Ok, let me rephrase: We should truncate the responses to whatever the client supports (512 without EDNS, whatever is indicated with EDNS) and 65535 records for TCP.

**Comment by tsenart** [comment]
:+1:

**Comment by sargun** [comment]
@discordianfish Is this what your PR that you merged today does?

**Comment by jdef** [comment]
#330 (merged today) should take care of proper truncation. is there a need
for a separate follow-up ticket re: the "randomly load balance" part?

**Comment by jdef** [comment]
if we're already shuffling then it sounds like we're all set with this
ticket.

**Comment by sargun** [comment]
:smile_cat:
